### PR TITLE
Initial work for extension process hosted server.

### PR DESCRIPTION
**Pull request by dlebu**
Sending out this PR for design review. Please take a look and let me know if the current infrastructure supports all our future scenarios.

@MSLaguana @meenakb @digeff @MShamiltonia

### CODE REVIEW

File: src/common/extensionMessaging.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for details.
+
+import * as Q from "q";
+import * as http from "http";
+
+/**
+ * Message server parameters.
+ */
+export let ServerParams = {
+    PORT: 8099,
```
Comment (MSLaguana) [inline src/common/extensionMessaging.ts]: The port should be configurable, but that configuration will need to be communicated from the extension process to the debug adapter somehow. I suppose in the worst case we can ask the user to specify it in two locations: settings.json and launch.json?

Comment (dlebu) [inline src/common/extensionMessaging.ts]: Can't the debugger process parse `settings.json` or the debugger process parse `launch.json` programmatically?

File: src/extension/rn-extension.ts [hunk 2]
```diff
@@ -49,6 +50,10 @@ export function activate(context: vscode.ExtensionContext): void {
                     const nodeDebugPath = vscode.extensions.getExtension("andreweinand.node-debug").extensionPath;
                     const fsUtil = new FileSystem();
                     fsUtil.writeFile(path.resolve(__dirname, "../", "debugger", "nodeDebugLocation.json"), JSON.stringify({ nodeDebugPath })).done();
+                }).then(() => {
+                    let server = new ExtensionServer();
+                    server.setup();
```
Comment (MSLaguana) [inline src/extension/rn-extension.ts]: I see that `setup` returns a promise, but you don't care about the return value here. Should we be waiting until the server configures successfully or otherwise informing the user that something has gone wrong?

File: src/common/extensionMessaging.ts (hunk 1 above)
Comment (MSLaguana) [inline src/common/extensionMessaging.ts]: The debugger process could try and parse settings.json in the project, but the settings as seen by the extension is a combination of a global settings.json and the project's particular settings.json file, and replicating that logic could be messy.

We could try to have the extension parse the launch.json since we would already require that they all have matching ports (otherwise some configurations would fail) so that might be workable.

We also need to make sure that the server is restartable (since it's registered with the context that might be enough) so that if the settings/launch.json are modified by the user we update the port the server is listening on.

File: src/extension/rn-extension.ts (hunk 2 above)
Comment (digeff) [inline src/extension/rn-extension.ts]: I'm rewriting how we handle and report the errors at the top level, so you can leave the code as-is and add a TODO, and I'll implement it as part of my fix/PR.

File: src/extension/extensionServer.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        return deferred.promise;
+    }
+
+    /**
+     * Stops the server.
+     */
+    public dispose(): void {
+        if (this.serverInstance) {
+            this.serverInstance.close();
+        }
```
Comment (digeff) [inline src/extension/extensionServer.ts]: I'd assign null to serverInstance after we close it, just to have 2 possible states in the object (null or non-null and opened).

File: src/extension/extensionServer.ts [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+        message.on("end", () => {
+            let extensionMessage: em.MessageWithArgs<em.ExtensionIncomingMessage> = JSON.parse(body);
+            console.log("Received message: " + extensionMessage.message);
+            this.handleExtensionMessage(extensionMessage)
+                .then(result => {
+                    response.writeHead(200, "OK", { "Content-Type": "application/json" });
+                    response.end(JSON.stringify(result));
+                })
+                .done();
```
Comment (digeff) [inline src/extension/extensionServer.ts]: return a 404 on rejection?

File: src/extension/extensionServer.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        });
+    }
+
+    /**
+     * Extension message handler.
+     */
+    private handleExtensionMessage(messageWithArgs: em.MessageWithArgs<em.ExtensionIncomingMessage>): Q.Promise<any> {
+        let deferred = Q.defer<any>();
+        /* handle each extension message here and return the result */
+        switch (messageWithArgs.message) {
```
Comment (digeff) [inline src/extension/extensionServer.ts]: If we are planning on supporting a lot of messages, I'd use a dictionary here instead of a switch.

File: src/common/extensionMessaging.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    START_PACKAGER,
+    STOP_PACKAGER
+}
+
+/**
+ * Generic interface for messages with arguments.
+ */
+export interface MessageWithArgs<T> {
+    message: T;
+    args?: any[];
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: Instead of using an any, I think I'd be better if we defined an interface for each message type we are planning on sending , and then the type of the message is MessageType1 | MessageType2 | MessageType3 etc... (Initially we'd just need a single message type to handle executing commands from the command pallete)

File: src/common/extensionMessaging.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    PORT: 8099,
+    HOST: "127.0.0.1"
+};
+
+/**
+ * Defines the messages sent to the extension.
+ * Add new messages to this enum.
+ */
+export enum ExtensionIncomingMessage {
+    START_PACKAGER,
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: I'd go with an ExecuteCommand message with a "startPackager", "stopPackager" argument, that can potentially be "runIos" or "runAndroid" too for free if we need to (given that we already have those methods on the commandPaletteHandler). Another alternative to consider, is that we could instead just call a VS Code API to execute a command ( https://code.visualstudio.com/Docs/extensionAPI/vscode-api#commands.executeCommand ). I think that our commands should appear there, so we could potentially use that to reapply breakpoints too when they release it, or any future commands that we add.

File: src/common/extensionMessaging.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    STOP_PACKAGER
+}
+
+/**
+ * Generic interface for messages with arguments.
+ */
+export interface MessageWithArgs<T> {
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: I'd put the classes and stuff shared between the client and the server in a common third file like ExtensionCommunicationProtocol.ts

File: src/common/extensionMessaging.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+ */
+export class ExtensionMessageSender {
+
+    public sendMessage(message: MessageWithArgs<ExtensionIncomingMessage>): Q.Promise<any> {
+        let deferred = Q.defer<any>();
+
+        let options = {
+            host: ServerParams.HOST,
+            port: ServerParams.PORT,
+            path: "/",
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: I'd put the type of the message in the path for example:
path: /executeCommand
body: { commandName: "startPackager" }

If you like RESTful you can even go
/commands/execute

File: src/extension/extensionServer.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            case em.ExtensionIncomingMessage.START_PACKAGER:
+                /* TODO */
```
Comment (digeff) [inline src/extension/extensionServer.ts]: We might not need this for V1, and maybe we can put it off for now, but in the long term I think we should expose the async nature of the operations in the server as in:
/commands/execute { command: "startPackager" } ---> Returns ID 1234
/commands/queryExecutionStatus { id: "1234" } ---> Running
/commands/queryExecutionStatus { id: "1234" } ---> Running
/commands/queryExecutionStatus { id: "1234" } ---> Running
/commands/queryExecutionStatus { id: "1234" } ---> Finished

File: src/common/extensionMessaging.ts (hunk 1 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: Updating to read from launch.json.

File: src/extension/rn-extension.ts (hunk 2 above)
Comment (dlebu) [inline src/extension/rn-extension.ts]: Adding TODO.

File: src/extension/extensionServer.ts (hunk 3 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: Agreed.

File: src/extension/extensionServer.ts (hunk 5 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: I don't see a lot of messages yet. If we reach a point where this is becoming unmanageable, we can move to a dictionary.

File: src/common/extensionMessaging.ts (hunk 6 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: Keeping in mind that we will probably not have a lot of messages, and most of them without arguments, do you still think we need to create an interface for each method type?

File: src/common/extensionMessaging.ts (hunk 7 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: Again I think this is overengineering for this feature. @MSLaguana @meenakb what do you guys think?

File: src/extension/extensionServer.ts (hunk 10 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: What would be the utility of polling? Where do you see this information being used in the product? We can certainly display progress with the current design as well - the server will not return until the operation is not finished.

File: src/common/extensionMessaging.ts (hunk 9 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: @MSLaguana @joshuaskelton @guillaumejenkins @meenakb Thoughts?

File: src/extension/extensionServer.ts (hunk 4 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: I like that.

File: src/common/extensionMessaging.ts (hunk 8 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: The `extensionMessaging.ts` file is common and does not depend on `vscode` or the debugging environment. Do you mean to move the `ExtensionMessageSender` class outside?

File: src/common/extensionMessaging.ts (hunk 9 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: I like REST style APIs but I think this is not the place to use them. If we do this we will end up putting even more logic in the server and the communication protocol (building paths, etc). What are the advantages do you think we will get by using separate paths for messages?

File: src/common/extensionMessaging.ts (hunk 7 above)
Comment (MSLaguana) [inline src/common/extensionMessaging.ts]: Considering we own the client and server components and they will be updated in-sync, I am inclined to keep it simple for now and only add complexity when we have a better idea of what we need. We don't have to worry about backwards compatibility so refactoring later won't be much trouble.

File: src/common/extensionMessaging.ts (hunk 9 above)
Comment (MSLaguana) [inline src/common/extensionMessaging.ts]: My first instinct is that we should pick an approach and go with it: Either a socket protocol of sorts with particular formatted messages on the wire, or a HTTP protocol where we use URLs and the body of the request is minimal.

**Comment by dlebu** [comment]
I met with @MSLaguana and @digeff yesterday and we decided on the following:

- we send messages to the server by POSTing to `http://localhost:PORT/MESSAGE_NAME`
- extra parameters will be sent in the request body
- the server port will be configurable in `launch.json`

The changes are in, please take a look.

File: src/extension/extensionServer.ts (hunk 5 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: I added a dictionary.

File: package.json [hunk 11]
```diff
@@ -60,6 +60,7 @@
                         "request": "launch",
                         "platform": "android",
                         "internalDebuggerPort": 9090,
+                        "extensionServerPort": 8099,
```
Comment (digeff) [inline package.json]: What about internalExtensionPort to be consistent with the other port?

File: src/extension/settingsHelper.ts [hunk 12]
```diff
@@ -63,4 +67,22 @@ export class SettingsHelper {
...
     }
+
+    /**
+     * Reads and parses the launch.json file.
+     */
+    public static readLaunchJson(): Q.Promise<any> {
+        let fileSystem = new FileSystem();
+        let launchJsonPath = SettingsHelper.launchJsonPath;
+
+        return fileSystem.exists(launchJsonPath)
```
Comment (digeff) [inline src/extension/settingsHelper.ts]: I think you should just open the readFile, without checking if it exists or not

File: src/extension/settingsHelper.ts [hunk 13]
```diff
@@ -63,4 +67,22 @@ export class SettingsHelper {
...
+            .then(exists => {
+                if (exists) {
+                    return fileSystem.readFile(launchJsonPath)
+                        .then(content => { return JSON.parse(content); });
+                } else {
+                    return null;
```
Comment (digeff) [inline src/extension/settingsHelper.ts]: I'd reject the promise instead of returning null

File: src/common/extensionMessaging.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    STOP_PACKAGER,
+    PREWARM_BUNDLE_CACHE
+}
+
+/**
+ * Sends messages to the extension.
+ */
+export class ExtensionMessageSender {
+
+    public sendMessage(message: ExtensionMessage, args?: any[], port?: number): Q.Promise<any> {
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: I think we should send the parameters by name using args: any instead of by position using args: any[]

File: src/debugger/scriptImporter.ts [hunk 15]
```diff
@@ -36,28 +37,37 @@ export class ScriptImporter {
...
                     return { contents: scriptBody, filepath: scriptFilePath };
                 }).finally(() => {
                     // Request that the debug adapter update breakpoints and sourcemaps now that we have written them
-                    return new Request().request(`http://localhost:${this.debugAdapterPort}/refreshBreakpoints`);
+                    return new Request().request(`http://localhost:${debugAdapterPort}/refreshBreakpoints`);
                 });
         });
     }
 
+    public downloadDebuggerWorker(sourcesStoragePath: string): Q.Promise<void> {
```
Comment (digeff) [inline src/debugger/scriptImporter.ts]: I wouldn't put downloadDebuggerWorker here. ScriptImporter was cohesive before just being responsible for downloading the scripts that the RN App requests, and this method is not related to that responsibility.

File: src/extension/extensionServer.ts [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        let launchCallback = (error: any) => {
+            if (error) {
+                deferred.reject(error);
+            } else {
+                deferred.resolve(null);
+            }
+        };
+
+        let port = em.ServerDefaultParams.PORT;
+        SettingsHelper.readLaunchJson()
```
Comment (digeff) [inline src/extension/extensionServer.ts]: Can you refactor the file reading to another method and pass it as a parameter? That will make is easier when we write tests, and we want to just pass the configuration without reading the file.

File: src/common/extensionMessaging.ts (hunk 14 above)
Comment (digeff) [inline src/common/extensionMessaging.ts]: I'd also make args be something like args1|args2|args3 so we get some kind of typing for this, instead of using any

File: src/common/extensionMessaging.ts [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+
+        let responseCallback = (response: http.IncomingMessage) => {
+            let body = "";
+
+            response.on("data", function(data: any) {
+                body += data;
+            });
+
+            response.on("end", function() {
+                let responseBody: any = body ? JSON.parse(body) : null;
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: An error while parsing should reject the deferred

File: src/debugger/launcher.ts [hunk 18]
```diff
@@ -32,17 +34,22 @@ export class Launcher {
...
+                            return scriptImporter.downloadDebuggerWorker(sourcesStoragePath).then(() => {
+                                Log.logMessage("Downloaded debuggerWorker.js (Logic to run the React Native app) from the Packager.");
+                            });
                         })
                         // We've seen that if we don't prewarm the bundle cache, the app fails on the first attempt to connect to the debugger logic
                         // and the user needs to Reload JS manually. We prewarm it to prevent that issue
                         .then(() => {
                             generator.step("prewarmBundleCache");
-                            return packager.prewarmBundleCache(runOptions.platform);
+                            return extensionMessageSender.sendMessage(em.ExtensionMessage.PREWARM_BUNDLE_CACHE, [runOptions.platform]);
```
Comment (MSLaguana) [inline src/debugger/launcher.ts]: You still need to add in the ability to pass on the port read from launch.json, right?

File: package.json (hunk 11 above)
Comment (dlebu) [inline package.json]: Renamed. We add them here so that it's easier for users to change the default values. If no value is specified, the default is used.

File: src/common/launchArgs.ts [hunk 19]
```diff
@@ -12,8 +12,12 @@ export interface IRunOptions extends ILaunchArgs {
...
  */
-interface ILaunchArgs {
+export interface ILaunchArgs {
     platform?: string;
     target?: string;
     debugAdapterPort?: number;
+    extensionServerPort?: number;
+    /* internal properties */
+    internalDebuggerPort?: any;
+    args?: string[];
```
Comment (MSLaguana) [inline src/common/launchArgs.ts]: This is currently conflating 2 different launch args: The original ILaunchArgs here was arguments-as-interpreted-by-debugee, and was constructed based on command line arguments. The new interpretation adds in arguments-as-specified-in-launch.json which is slightly different.

For example the `args` property is the nodejs debugger args property which we use to pass arguments to the debugee process. I think that we should avoid mashing the two together because they are used in different ways.

File: package.json (hunk 11 above)
Comment (MSLaguana) [inline package.json]: Note that the "default" specified here is only for documentation. We need to ensure that our code copes with missing values and explicitly set a default.

File: src/debugger/launcher.ts (hunk 18 above)
Comment (MSLaguana) [inline src/debugger/launcher.ts]: Oh, You have added in the pass through, it just isn't used here or further above yet

File: src/extension/settingsHelper.ts (hunk 12 above)
Comment (dlebu) [inline src/extension/settingsHelper.ts]: Since we are aiming at returning a JSON object, I believe returning `null` if the file is not there is a good contract for consumers of this method.

File: src/extension/settingsHelper.ts (hunk 13 above)
Comment (dlebu) [inline src/extension/settingsHelper.ts]: See comment above.

File: src/extension/extensionServer.ts (hunk 16 above)
Comment (MSLaguana) [inline src/extension/extensionServer.ts]: I believe that we also need a file watcher or to hook onto some vscode event if there is one for launch.json updating. If the user updates the launch.json we should restart the server respecting the new value

File: src/extension/settingsHelper.ts (hunk 12 above)
Comment (digeff) [inline src/extension/settingsHelper.ts]: Given that for the purpose of this method, the file not being there, or being unable to read it for any reason is the same, I'd argue that we should either return the JSON object, with a resolved promise. And if for any reason we cna't do that, we should just reject the promise and let the caller handle it somehow. Adding a third possible case that we tread identically to the second one, just adds unnecessary complexity.

File: src/extension/extensionServer.ts [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    private handleIncomingMessage(message: http.IncomingMessage, response: http.ServerResponse): void {
+        let body = "";
+        message.on("data", (chunk: string) => {
+            body += chunk;
+        });
+
+        message.on("end", () => {
+            let args: any[];
+            if (body) {
+                args = JSON.parse(body);
```
Comment (MSLaguana) [inline src/extension/extensionServer.ts]: try/catch with a 404 on invalid JSON?

File: src/debugger/launcher.ts (hunk 18 above)
Comment (MSLaguana) [inline src/debugger/launcher.ts]: On closer inspection, you have added code below to receive the port, but not yet added the code to nodeDebugWrapper to send it through. See https://github.com/Microsoft/vscode-react-native/pull/76/files#diff-933d1a6506e93a563f1ed6dc80ff0f47R126

File: package.json (hunk 11 above)
Comment (digeff) [inline package.json]: The thing is that we are polluting the UI with things he shouldn't really care about. It would be nice if we could magically solve this problem under the rug.

File: src/common/extensionMessaging.ts (hunk 14 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: I updated the handler type to take parameters instead of an array. This way, whatever you send to the server should match the arguments on the message handler.

File: src/debugger/scriptImporter.ts (hunk 15 above)
Comment (dlebu) [inline src/debugger/scriptImporter.ts]: I disagree. The `ScriptImporter` is responsible with downloading scripts from the packager and saving them to disk, no matter what they are used for. It is a much better place for this logic than the `Packager` class, where I moved it from.

File: src/extension/extensionServer.ts (hunk 16 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: Wouldn't it be easier to mock the `SettingsHelper` in tests if we want to do that?

File: src/common/extensionMessaging.ts (hunk 17 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: Agreed.

File: src/debugger/launcher.ts (hunk 18 above)
Comment (dlebu) [inline src/debugger/launcher.ts]: Fixed.

File: src/extension/extensionServer.ts (hunk 20 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: Fixed.

File: src/extension/settingsHelper.ts (hunk 12 above)
Comment (dlebu) [inline src/extension/settingsHelper.ts]: Fixed.

File: src/debugger/nodeDebugWrapper.ts [hunk 21]
```diff
@@ -132,6 +133,7 @@ Telemetry.init("react-native-debug-adapter", version, true).then(() => {
         args.args = [
             args.platform,
             debugServerListeningPort.toString(),
+            args.internalExtensionPort.toString(),
```
Comment (MSLaguana) [inline src/debugger/nodeDebugWrapper.ts]: I would suggest also adding in default-setting logic here to avoid any issues with unexpected input since it is user-exposed. That would also prepare it for when we move the bulk of the initialisation code to the debug adapter and so the value would be consumed here as well.

File: src/debugger/nodeDebugWrapper.ts [hunk 22]
```diff
@@ -43,6 +43,7 @@ interface ILaunchArgs {
     platform: string;
     target?: string;
     internalDebuggerPort?: any;
```
Comment (digeff) [inline src/debugger/nodeDebugWrapper.ts]: why is this any?

Comment (dlebu) [inline src/debugger/nodeDebugWrapper.ts]: @MSLaguana ?

Comment (MSLaguana) [inline src/debugger/nodeDebugWrapper.ts]: I did it this way because the actual value/type is user provided. Initially I had said that it was a number, but then I couldn't `parseInt` even though it works fine at runtime to ensure that we end up with a number.

File: src/extension/extensionServer.ts (hunk 16 above)
Comment (digeff) [inline src/extension/extensionServer.ts]: I don't think so. If you refactor it you'll just need to call setup(1234); if you don't you'll need to create a mock SettingsHelper.readLaunchJson, install it, etc...

File: src/debugger/nodeDebugWrapper.ts (hunk 22 above)
Comment (digeff) [inline src/debugger/nodeDebugWrapper.ts]: If VS Code doesn't guarantee any kind of type validation, then we should mark all the parameters as any, and make sure that we handle them if we are not what we expect them to be.

File: src/debugger/nodeDebugWrapper.ts (hunk 21 above)
Comment (dlebu) [inline src/debugger/nodeDebugWrapper.ts]: Added.

File: src/debugger/nodeDebugWrapper.ts (hunk 22 above)
Comment (dlebu) [inline src/debugger/nodeDebugWrapper.ts]: @digeff This is a user provided value. We make the best effort to parse it, whether it is a string or a number. We used `any` because compilation fails if you pass anything other than string to parse.

File: src/extension/extensionServer.ts (hunk 16 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: @MSLaguana Based on our discussion, I added a comment in package.json to reflect the need of restarting VSCode if the port changes.

File: src/debugger/nodeDebugWrapper.ts (hunk 22 above)
Comment (digeff) [inline src/debugger/nodeDebugWrapper.ts]: I agree with the any. I'm saying that all the other user provided values should also be any, and they should be marked as string or number only after we've successfully parsed them or validate that they are the proper type.

**Comment by MSLaguana** [comment]
I've noticed a couple of incorrectly cased import statements, but apart from that it looks good to me

File: src/extension/extensionServer.ts [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+            } else {
+                deferred.resolve(null);
+            }
+        };
+
+        let port = em.ServerDefaultParams.PORT;
+        SettingsHelper.readLaunchJson()
+            .then((launchJson: any) => {
+                if (launchJson) {
+                    /* take the fist configuration that specifies the port */
```
Comment (digeff) [inline src/extension/extensionServer.ts]: What happens if the user has different configurations with different internal ports? Maybe he uses one port for Android and one for iOS. Maybe he uses both Windows and Mac and has Run Android on Windows and Run Android on iOS and passes a different port to each one.

File: src/extension/settingsHelper.ts [hunk 24]
```diff
@@ -63,4 +67,16 @@ export class SettingsHelper {
                 return settingsJson["typescript.tsdk"] || "";
             });
     }
+
+    /**
+     * Reads and parses the launch.json file.
+     */
+    public static readLaunchJson(): Q.Promise<any> {
```
Comment (digeff) [inline src/extension/settingsHelper.ts]: We read JSON files in a few different places. I'd refactor this to take the file path as a parameter, and put a generic readJson method in the FileSystem class.

File: src/extension/extensionServer.ts (hunk 23 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: I added a comment about this in `package.json` to make sure the values are in sync. For your described scenario, they would need to find a common internal port that is available.

**Comment by MSLaguana** [comment]
I think this change to use named pipes is much better, and the implementation looks good from what I've seen so far. If it works on windows, it seems good to me.

File: src/common/extensionMessaging.ts [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+import * as Q from "q";
+import * as net from "net";
+
+/**
+ * Pipe path used for communicating with the server.
+ */
+let WIN_ServerPipePath = "\\\\?\\pipe\\vscodereactnative";
+let UNIX_ServerPipePath = "/tmp/vscodereactnative.sock";
+
+export let getPipePath = (): string => {
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: I'd put this method inside a class. Process can't be mocked with Mockery, so the cleanest option for testing this or dependent methods will probably be to pass a mock process as a constructor argument to some class.

File: src/common/extensionMessaging.ts [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    message: ExtensionMessage;
+    args: any[];
+}
+
+/**
+ * Sends messages to the extension.
+ */
+export class ExtensionMessageSender {
+
+    public sendMessage(message: ExtensionMessage, args?: any[]): Q.Promise<any> {
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: I think that changing this method to expect an already built MessageWithArguments  instead of 2 parameters will make the code simpler.

File: src/common/extensionMessaging.ts [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+        let deferred = Q.defer<any>();
```
Comment (digeff) [inline src/common/extensionMessaging.ts]: Given that no-one seems to be using the return value at the moment, I'd change this method to return Q.Promise<void> instead of Q.Promise<any>.

File: src/extension/extensionServer.ts [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+import {FileSystem} from "../common/node/fileSystem";
+import {Packager} from "../common/packager";
+import {Log} from "../common/log";
+import * as Q from "q";
+import * as net from "net";
+import * as vscode from "vscode";
+
+
+export class ExtensionServer implements vscode.Disposable {
+    private outputChannel: any;
```
Comment (digeff) [inline src/extension/extensionServer.ts]: Why is this any?

File: src/extension/extensionServer.ts [hunk 29]
```diff
@@ -0,0 +1,141 @@
...
+
+    /**
+     * Starts the server.
+     */
+    public setup(): Q.Promise<void> {
+
+        let deferred = Q.defer<void>();
+
+        let launchCallback = (error: any) => {
+            Log.logMessage("Server started." + (error ? error : ""));
```
Comment (digeff) [inline src/extension/extensionServer.ts]: We should call logError if you have an error. Should we print a more detailed message to the user if he gets an error here? He probably won't be able to use the debugging functionality.

File: src/extension/extensionServer.ts [hunk 30]
```diff
@@ -0,0 +1,141 @@
...
+        return this.reactNativePackager.prewarmBundleCache(platform);
+    }
+
+    /**
+     * Extension message handler.
+     */
+    private handleExtensionMessage(messageWithArgs: em.MessageWithArguments): Q.Promise<any> {
+        let handler = this.messageHandlerDictionary[messageWithArgs.message];
+        if (handler) {
+            Log.logMessage("Handling message: " + em.ExtensionMessage[messageWithArgs.message]);
```
Comment (digeff) [inline src/extension/extensionServer.ts]: If this is intended to be an internal message, we should call logInternalMessage instead

File: src/extension/extensionServer.ts [hunk 31]
```diff
@@ -0,0 +1,141 @@
...
+        } else {
+            return Q.reject("Invalid message: " + messageWithArgs.message);
+        }
+    }
+
+    /**
+     * Handles connections to the server.
+     */
+    private handleSocket(socket: net.Socket): void {
+        let dataCallback = (data: any) => {
```
Comment (digeff) [inline src/extension/extensionServer.ts]: I'd remove this variable and just declare the function inside line 114.

File: src/extension/extensionServer.ts [hunk 32]
```diff
@@ -0,0 +1,141 @@
...
+            try {
+                let messageWithArgs: em.MessageWithArguments = JSON.parse(data);
+                this.handleExtensionMessage(messageWithArgs)
+                    .then(result => {
+                        socket.end(JSON.stringify(result));
+                    })
+                    .catch(() => { socket.end(); })
```
Comment (digeff) [inline src/extension/extensionServer.ts]: Should we return some kind of 404 or error to the socket?

File: src/common/extensionMessaging.ts (hunk 27 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: Not sure what you mean here. `Q.Promise` instead of `Q.Promise`?

Comment (digeff) [inline src/common/extensionMessaging.ts]: Git is interpreting it as markup. There is a <void> and a <any> after those two promises :)

File: src/extension/extensionServer.ts [hunk 33]
```diff
@@ -0,0 +1,141 @@
...
+    };
+
+    /**
+     * Recovers the server in case the named socket we use already exists, but no other instance of VSCode is active.
+     */
+    private recoverServer(error: any): void {
+        let errorHandler = (e: any) => {
+            /* The named socket is not used. */
+            if (e.code === "ECONNREFUSED") {
+                new FileSystem().removePathRecursivelyAsync(em.getPipePath())
```
Comment (digeff) [inline src/extension/extensionServer.ts]: Could you please add some comments explaining why we are using removePathRecursivelyAsync here instead of removing a single file or directory?

File: src/common/extensionMessaging.ts (hunk 27 above)
Comment (dlebu) [inline src/common/extensionMessaging.ts]: I disagree, we should be able to return values here. We wouldn't want to modify the communication protocol later.

File: src/extension/extensionServer.ts (hunk 28 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: Chaning.

**Comment by digeff** [comment]
:+1: with some comments

File: src/extension/extensionServer.ts (hunk 33 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: That is what `removePathRecursivelyAsync` does internally - removes a directory or a file, if they exist.

File: src/extension/extensionServer.ts (hunk 32 above)
Comment (dlebu) [inline src/extension/extensionServer.ts]: Good idea, I'll introduce an error marker.

File: src/common/extensionMessaging.ts (hunk 27 above)
Comment (MSLaguana) [inline src/common/extensionMessaging.ts]: I'm inclined to agree with Daniel. Potentially we could make the return type an `IMessageResponse` type which is an opaque `{}` type, but I think `any` is just about as good.

**Comment by dlebu** [comment]
Thanks everyone. Merging to master.
